### PR TITLE
Terra meerkat

### PR DESCRIPTION
**Pull request by Chris-green-stfc**
Moved VM creation / deletion from ansible to terraform, for better error handling.
The storage benchmarking now points to a victoria metrics DB.

### CODE REVIEW

File: AnsiblePlaybooks/meerkat/deploy.tf [hunk 1]
```diff
@@ -0,0 +1,42 @@
+terraform {
+			required_version = ">= 0.14.0"
+			  required_providers {
+			    openstack = {
+			      source  = "terraform-provider-openstack/openstack"
+			      version = "~> 1.53.0"
+			    }
+			}
+}
```
Comment (anish-mudaraddi) [inline AnsiblePlaybooks/meerkat/deploy.tf:9]: some strange formatting here - please check

File: AnsiblePlaybooks/meerkat/roles/storage_benchmark/files/benchmark.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	echo Seq read: $sys_seqr_1 MB/s
+	echo Rand write: $sys_randw_1 MB/s
+	echo Rand read: $sys_randr_1 MB/s
+fi
+
+########################################################################################################################
+# 1000 x 1 MB
+echo -------------------------------------------------------
+n=1000
+s=1
```
[the remark this answers is not in the document]
Comment (anish-mudaraddi) [reply AnsiblePlaybooks/meerkat/roles/storage_benchmark/files/benchmark.sh:31]: reading further into this - it looks like you run 3 benchmarking scripts in one here - see my comment on splitting it up

File: AnsiblePlaybooks/meerkat/roles/storage_benchmark/files/benchmark.sh [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+server_UUID=$(curl  http://169.254.169.254/openstack/latest/meta_data.json -s  | grep -ioP .{8}-.{4}-.{4}-.{4}-.{12})
+vm_info=$(openstack server show $server_UUID --os-cloud openstack)
+#hypervisor=$(echo $vm_info | grep -ioP "\w*.nubes.rl.ac.uk" | head -1)
+flavor=$(echo $vm_info | grep -ioP "flavor\s\|\s\S*" | cut -c 10-)
+image=$(echo $vm_info | grep -ioP "image\s\|\s\S*" | cut -c 9-)
+echo $flavor
+echo $image
+
+#KB results
+curl -H 'Content-Type: application/json' -d '{"metric":"sequential_write","value":"'$sys_seqw_1'","tags":{"flavor":"'$flavor'", "image":"'$image'","test_type":"kb", "storage_type":"local"}}' http://172.16.101.186:4242/api/put
```
[the remark this answers is not in the document]
Comment (Chris-green-stfc) [reply AnsiblePlaybooks/meerkat/roles/storage_benchmark/files/benchmark.sh]: This is an openstack IP address, so will never change

File: AnsiblePlaybooks/meerkat/roles/storage_benchmark/files/benchmark.sh [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+sys_seqr_3=$(echo "$output" | grep -ioP "Sequential read: +(?)\d+(?:\.\d+)?" | grep -oP "\d+(?:\.\d+)?")
+sys_randw_3=$(echo "$output" | grep -ioP "Random write: +(?)\d+(?:\.\d+)?" | grep -oP "\d+(?:\.\d+)?")
+sys_randr_3=$(echo "$output" | grep -ioP "Random read: +(?)\d+(?:\.\d+)?" | grep -oP "\d+(?:\.\d+)?")
+echo Seq write: $sys_seqw_3 MB/s
+echo Seq read: $sys_seqr_3 MB/s
+echo Rand write: $sys_randw_3 MB/s
+echo Rand read: $sys_randr_3 MB/s
+
+
+server_UUID=$(curl  http://169.254.169.254/openstack/latest/meta_data.json -s  | grep -ioP .{8}-.{4}-.{4}-.{4}-.{12})
```
[the remark this answers is not in the document]
Comment (Chris-green-stfc) [reply AnsiblePlaybooks/meerkat/roles/storage_benchmark/files/benchmark.sh]: This is an openstack constant IP address, so will never change

File: AnsiblePlaybooks/meerkat/roles/storage_benchmark/files/benchmark.sh [hunk 5]
```diff
@@ -0,0 +1,106 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (Chris-green-stfc) [reply AnsiblePlaybooks/meerkat/roles/storage_benchmark/files/benchmark.sh]: I have combined the benchmarks into one function, that is called 3 times. This way it is easier to add/alter tests

File: AnsiblePlaybooks/meerkat/meerkat.yaml [hunk 6]
```diff
@@ -15,7 +15,10 @@
     - role: update_machine
       tags: [ always ]
 
-    - role: storage_benchmark
+    - role: setup_torage_benchmark
```
Comment (anish-mudaraddi) [inline AnsiblePlaybooks/meerkat/meerkat.yaml:18]: typo

File: AnsiblePlaybooks/meerkat/roles/setup_storage_benchmark/files/benchmark.sh [hunk 7]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
```
Comment (anish-mudaraddi) [inline AnsiblePlaybooks/meerkat/roles/setup_storage_benchmark/files/benchmark.sh:1]: This is a lot better - great work :+1:

File: AnsiblePlaybooks/meerkat/roles/setup_storage_benchmark/files/benchmark.sh [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+# Get VM infor
+server_UUID=$(curl  http://169.254.169.254/openstack/latest/meta_data.json -s  | grep -ioP .{8}-.{4}-.{4}-.{4}-.{12})
+vm_info=$(openstack server show $server_UUID --os-cloud openstack)
+#hypervisor=$(echo $vm_info | grep -ioP "\w*.nubes.rl.ac.uk" | head -1)
+flavor=$(echo $vm_info | grep -ioP "flavor\s\|\s\S*" | cut -c 10-)
+image=$(echo $vm_info | grep -ioP "image\s\|\s\S*" | cut -c 9-)
+
+DB="http://172.16.101.186:4242/api/put"
```
Comment (anish-mudaraddi) [inline AnsiblePlaybooks/meerkat/roles/setup_storage_benchmark/files/benchmark.sh:49]: I'm happy to leave this as is for now - but in future you shouldn't hardcode the IPs. 

We should instead get it as an input when calling the script like so: 
```
# start of script
DB_IP=$1  #first argument to benchmark script
DB_PORT=$2 #second argument to benchmark script
...
DB="http://${DB_IP}:${DB_PORT}/api/put"
```

Makes it easier to change later - we can have them be a variable in Ansible like so;
```
- name: Run benchmark
  shell: ./benchmark.sh {{ db_ip }} {{ db_port }}
  register: script_output
```

You can leave this for a future PR when we need to change it

**Review by anish-mudaraddi — CHANGES_REQUESTED** [review]
looks good - you've got a typo :)

File: AnsiblePlaybooks/meerkat/roles/setup_storage_benchmark/files/benchmark.sh (hunk 8 above)
Comment (Chris-green-stfc) [reply AnsiblePlaybooks/meerkat/roles/setup_storage_benchmark/files/benchmark.sh:49]: This IP address will change. I plan to add in the variables like this to my next PR, with templating files.

**Review by anish-mudaraddi — APPROVED** [review]

**Pull request merged**